### PR TITLE
Implement auto sending requests

### PR DESCRIPTION
**Pull request by WaffleLapkin**

### CODE REVIEW

**Comment by Hirrolot** [comment]
I'm worried about the technical debt that this PR would introduce because it comprises of lots of tricks and quite unintuitive control flow, despite of being working.

How about using a macro instead?

[[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=717a85d229cd3814977c6509395be556)]
```rust
macro_rules! send {
    ($req:expr) => {
        $req.send().await
    };
}

// Our hypothetical request.
struct Request;

impl Request {
    async fn send(self) -> i32 {
        123
    }
}

fn main() {
    let _123 = async {
        let req = Request;
        send!(req)
    };
}
```

**Comment by Hirrolot** [comment]
Also related to: https://github.com/rust-lang/rust/issues/67982.

**Comment by WaffleLapkin** [comment]
I don't think that macro would accomplish much.

1. Macros are a bit scary. Personally, I expect that a macro will do something complicated and not just `.await.send()`.
2. `send!(...)` vs `....send().await` not much a difference (I would prefer the later...)

I also don't think that this code is too complicated. It just does simple bot-wrapper-things + implements `Future` for its request. Indeed this requires some minor tricks around `Pin<_>`, but they are ok, in my opinion.

**Comment by Hirrolot** [comment]
I agree that `send!(...)` is not much better than `... .send().await`, and moreover would necessitate deprecation of `send!` after the language issue above eventually resolves. In contrast, your PR plays nice with the future directions of `IntoFuture`.

Concerning the PR code. I was confused that it uses unreachable expressions, but others conceived me that it's okay for Rusty futures, so nevermind.

File: src/bot/auto_send.rs [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+///
+/// Requests returned by `<AutoSend<_> as `[`Requester`]`>` are [`Future`]s
+/// which means that you can simply `.await` them instead of using
+/// `.send().await`.
+///
+/// Notes:
+/// 1. This wrapper should be the most outer i.e.: `AutoSend<CacheMe<Bot>>`
+///    will automatically send requests, while `CacheMe<AutoSend<Bot>>` - won't.
+/// 2. After first call to `poll` on a request you will unable to access payload
+///    nor could you use [`send_ref`](Request::send_ref)
```
Comment (Hirrolot) [inline src/bot/auto_send.rs:21]:
```suggestion
/// 2. After first call to `poll` on a request you will unable to access payload
///    nor could you use [`send_ref`](Request::send_ref).
```

File: src/bot/auto_send.rs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    /// Creates new `AutoSend`.
+    ///
+    /// Note: it's recommended to use [`RequesterExt::auto_send`] instead.
+    ///
+    /// [`RequesterExt::auto_send`]: crate::requests::RequesterExt::auto_send
+    pub fn new(inner: B) -> AutoSend<B> {
+        Self { bot: inner }
+    }
+
+    /// Allows to access inner bot
```
Comment (Hirrolot) [inline src/bot/auto_send.rs:51]:
```suggestion
    /// Allows to access the inner bot.
```

File: src/bot/auto_send.rs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    pub fn inner(&self) -> &B {
+        &self.bot
+    }
+
+    /// Unwraps inner bot
```
Comment (Hirrolot) [inline src/bot/auto_send.rs:56]:
```suggestion

    /// Unwraps the inner bot.
```

File: src/bot/auto_send.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    pub fn new(inner: R) -> Self {
+        Self(Inner::Request(inner))
+    }
+}
+
+/// Data of the `AutoRequest` used to not expose variants (I wish there were
+/// private enum variants).
+#[pin_project::pin_project(project = InnerProj, project_replace = InnerRepl)]
+enum Inner<R: Request> {
+    /// Unsent modifiable request.
```
Comment (Hirrolot) [inline src/bot/auto_send.rs:83]:
```suggestion

    /// An unsent modifiable request.
```

File: src/bot/auto_send.rs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    Request(R),
+    /// Sent request.
```
Comment (Hirrolot) [inline src/bot/auto_send.rs:85]:
```suggestion

    /// A sent request.
```

File: src/bot/auto_send.rs [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+impl<R: Request> Future for AutoRequest<R> {
+    type Output = Result<Output<R>, R::Err>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this: Pin<&mut Inner<_>> = self.as_mut().project().0;
+
+        match this.as_mut().project() {
+            // Poll underling future
```
Comment (Hirrolot) [inline src/bot/auto_send.rs:143]:
```suggestion

            // Poll the underling future.
```

**Comment by WaffleLapkin** [comment]
Well, if `IntoFuture` will support `async`/`await` just as `IntoIterator`+`for` then wrapper introduced in this PR will be useless too. We'll just add `: IntoFuture` bound for `Request`.

File: src/bot/auto_send.rs [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+            InnerProj::Future(fut) => {
+                let res = futures::ready!(fut.poll(cx));
+                // We've got the result, so we set the state to done
```
Comment (Hirrolot) [inline src/bot/auto_send.rs:146]:
```suggestion
                // We've got the result, so we set the state to done.
```

File: src/bot/auto_send.rs [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+                this.set(Inner::Done);
+                Poll::Ready(res)
+            }
+            // This future is fused
```
Comment (Hirrolot) [inline src/bot/auto_send.rs:150]:
```suggestion

            // This future is fused.
```

File: src/bot/auto_send.rs [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+            InnerProj::Done => Poll::Pending,
+            // The `AutoRequest` future was polled for the first time after
+            // creation. We need to transform it into sent form by calling
+            // `R::send` and doing some magic around Pin.
+            InnerProj::Request(_) => {
+                // Replace `Request(_)` by `Done(_)` to obtain ownership over
+                // the former.
+                let inner = this.as_mut().project_replace(Inner::Done);
+                // Map Request(req) to `Future(req.send())`
```
Comment (Hirrolot) [inline src/bot/auto_send.rs:159]:
```suggestion
               // Map Request(req) to `Future(req.send())`.
```

File: src/bot/auto_send.rs [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+                let inner = match inner {
+                    InnerRepl::Request(req) => Inner::Future(req.send()),
+                    // Practically this is unreachable, because we've just checked for
+                    // both `Future(_)` and `Done` variants.
+                    InnerRepl::Future(_) | InnerRepl::Done => done_unreachable(),
+                };
+                // Set the resulting `Future(_)` back to pin
```
Comment (Hirrolot) [inline src/bot/auto_send.rs:166]:
```suggestion
                // Set the resulting `Future(_)` back to pin.
```

File: src/bot/auto_send.rs [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+                this.set(inner);
+
+                // poll self again, this time
```
Comment (Hirrolot) [inline src/bot/auto_send.rs:169]:
```suggestion
                // Poll `self` again, this time.
```

File: src/lib.rs [hunk 12]
```diff
@@ -15,8 +15,9 @@
 #[macro_use]
 mod local_macros; // internal helper macros
```
Comment (Hirrolot) [inline src/lib.rs:16]:
```suggestion
// The internal helper macros.
mod local_macros;
```

File: src/lib.rs [hunk 13]
```diff
@@ -15,8 +15,9 @@
...
 
+// FIXME(waffle): rethink modules, find a place for wrappers
```
Comment (Hirrolot) [inline src/lib.rs:18]:
```suggestion
// FIXME(waffle): rethink modules, find a place for wrappers.
```

**Review by Hirrolot — CHANGES_REQUESTED** [review]

**Comment by Hirrolot** [comment]
> Well, if IntoFuture will support async/await just as IntoIterator+for then wrapper introduced in this PR will be useless too. We'll just add : IntoFuture bound for Request.

I understand; I meant that it would not break backward compatibility in user's code (at least in most cases).

File: src/bot/auto_send.rs (hunk 11 above)
Comment (WaffleLapkin) [reply src/bot/auto_send.rs:169]: Oh, I think I forgot something. It should be `this time <...>`

Comment (Hirrolot) [reply src/bot/auto_send.rs:170]: I've updated it. It's alright now?

Comment (WaffleLapkin) [reply src/bot/auto_send.rs:170]: No-no. I mean `Poll self again, this time <...>`. Maybe something like ``Poll self again, this time another brunch will be executed, returning `Poll` ``

Comment (Hirrolot) [reply src/bot/auto_send.rs:170]: Now?

**Review by Hirrolot — APPROVED** [review]

**Pull request merged**